maggy/core/patching/modules.py

```diff
@@ -0,0 +1,162 @@
+#
+#   Copyright 2021 Logical Clocks AB
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+from __future__ import annotations
+
+from types import SimpleNamespace
+from typing import Type, Any
+
+from torch.nn import Module as TorchModule
+from torch.nn.parallel import DistributedDataParallel as TorchDistributedDataParallel
+
+from deepspeed.pipe import PipelineModule
+from deepspeed.runtime.engine import DeepSpeedEngine
+from fairscale.nn import FullyShardedDataParallel as FairscaleFullyShardedDataParallel
```

Comment:
these are not installed by default on a hopsworks cluster, can you add them to setup.py as extras. We should think about adding them to the base environment on hopsworks if there are not dependency clashes. Until then I am thinking if it might be a good idea to catch ImportErrors here, in case they are not installed, so users can still use DDP.